fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/HNSW.java

```diff
@@ -0,0 +1,2022 @@
+                    final Estimator estimator = quantizer.estimator();
+
+                    final NodeReferenceWithDistance entryState =
+                            new NodeReferenceWithDistance(entryNodeReference.getPrimaryKey(),
+                                    entryNodeReference.getVector(),
+                                    estimator.distance(transformedQueryVector, entryNodeReference.getVector()));
+
+                    final int entryLayer = entryNodeReference.getLayer();
+                    return forLoop(entryLayer, entryState,
+                            layer -> layer >= 0,
+                            layer -> layer - 1,
+                            (layer, previousNodeReference) -> {
+                                if (layer == 0) {
+                                    // entry data points to a node in layer 0 directly
+                                    return CompletableFuture.completedFuture(previousNodeReference);
+                                }
+
+                                final var storageAdapter = getStorageAdapterForLayer(layer);
+                                return greedySearchLayer(storageAdapter, readTransaction, storageTransform, estimator,
+                                        previousNodeReference, layer, transformedQueryVector);
+                            }, executor)
+                            .thenCompose(nodeReference -> {
+                                if (nodeReference == null) {
+                                    return CompletableFuture.completedFuture(null);
+                                }
+
+                                final var storageAdapter = getStorageAdapterForLayer(0);
+
+                                return searchLayer(storageAdapter, readTransaction, storageTransform, estimator,
+                                        ImmutableList.of(nodeReference), 0, efSearch, Maps.newConcurrentMap(),
+                                        transformedQueryVector)
+                                        .thenApply(searchResult -> {
+                                            // reverse the original queue
+                                            final TreeMultimap<Double, NodeReferenceAndNode<? extends NodeReference>> sortedTopK =
+                                                    TreeMultimap.create(Comparator.naturalOrder(),
+                                                            Comparator.comparing(nodeReferenceAndNode ->
+                                                                    nodeReferenceAndNode.getNode().getPrimaryKey()));
+
+                                            for (final NodeReferenceAndNode<?> nodeReferenceAndNode : searchResult) {
+                                                if (sortedTopK.size() < k || Objects.requireNonNull(sortedTopK.keySet().last()) >
+                                                        nodeReferenceAndNode.getNodeReferenceWithDistance().getDistance()) {
```

Comment:
Maybe I'm a little unclear on the details here, but doesn't `searchLayer` return results in the order needed here by `kNearestNeighbors`, and so all this needs to do is truncate the list to the top `k`?